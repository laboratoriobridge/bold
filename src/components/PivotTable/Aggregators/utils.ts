import { Aggregator, AggregatorEnum, AggregatorLabels } from './model'

export const getCountAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.COUNT,
  label: aggregatorLabels.count,
  value: undefined,
  keyDependent: false,
})

const getPercentageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.PERCENTAGE,
  label: aggregatorLabels.percentage,
  value: (values: number[], total: number): number => (values.reduce((prev, curr) => prev + curr, 0) * 100) / total,
  keyDependent: false,
  chain: [getCountAggregator(aggregatorLabels)],
  suffix: '%',
})

const getAverageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.AVERAGE,
  label: aggregatorLabels.average,
  value: (values: number[]): number => values.reduce((prev, curr) => prev + curr, 0) / values.length,
  keyDependent: true,
})

const getMaximumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MAXIMUM,
  label: aggregatorLabels.maximum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev > curr ? prev : curr)),
  keyDependent: true,
})

const getMinimumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MINIMUM,
  label: aggregatorLabels.minimum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev < curr ? prev : curr)),
  keyDependent: true,
})

export const getAggregators = (aggregatorLabels: AggregatorLabels) => {
  return [
    getCountAggregator(aggregatorLabels),
    getPercentageAggregator(aggregatorLabels),
    getAverageAggregator(aggregatorLabels),
    getMaximumAggregator(aggregatorLabels),
    getMinimumAggregator(aggregatorLabels),
  ]
}

export const getKeyNotDependentAggregators = (aggregators: Aggregator[]) => {
  return aggregators.filter((agg) => !agg.keyDependent)
}
