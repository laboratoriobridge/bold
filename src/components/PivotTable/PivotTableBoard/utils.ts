import { difference, intersection } from 'lodash'

import { BoardField, FieldFiltersByKey, FieldValuesByKey } from './model'

interface InitialKeyPositionsAndValues<T> {
  initialRowKeys: Array<keyof T>
  initialColumnKeys: Array<keyof T>
  initialAvailableKeys: Array<keyof T>
}

export function getInitialKeys<T>(
  keys: FieldValuesByKey<T>,
  initialFields?: Array<BoardField<T>>
): InitialKeyPositionsAndValues<T> {
  const rowKeys: Array<keyof T> = []
  const columnKeys: Array<keyof T> = []

  initialFields?.forEach(({ key, origin }) => {
    if (origin === 'row') {
      rowKeys.push(key)
    }
    if (origin === 'column') {
      columnKeys.push(key)
    }
  })
  const availableKeys: Array<keyof T> = difference(Array.from(keys.keys()), [...rowKeys, ...columnKeys])

  return {
    initialRowKeys: rowKeys,
    initialColumnKeys: columnKeys,
    initialAvailableKeys: availableKeys,
  }
}

export function initializeActiveFilters<T>(
  allFiltersByKey: FieldFiltersByKey<T>,
  keys: FieldValuesByKey<T>,
  initialFields?: Array<BoardField<T>>,
  setFilterState?: (filter: FieldFiltersByKey<T>) => void
): void {
  const activeFiltersByKey = new Map(allFiltersByKey)

  initialFields?.forEach(({ key, filters }) => {
    if (filters && filters.length > 0) {
      const currentFieldValues = keys.get(key)
      const validFilters = intersection(filters, currentFieldValues)
      activeFiltersByKey.set(key, new Set(validFilters))
    }
  })

  setFilterState?.(activeFiltersByKey)
}

export function handleTagFilterRemove<T extends object>(
  key: keyof T,
  value: string,
  filterState: FieldFiltersByKey<T>
): Set<string> {
  const values = Array.from(filterState.get(key) ?? new Set<string>())

  return new Set(values.filter((item) => item !== value))
}
