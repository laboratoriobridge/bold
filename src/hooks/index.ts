export * from './useBreakpoint'
export * from './useClickOutside'
export * from './useFormControl'
export * from './useIsMounted'
export * from './useRovingTabIndex'
export * from './useScrollPosition'
export * from './useToastMessages'
export * from './useTransition'
export * from './useWindowSize'
export * from './useIsOverflowing'
