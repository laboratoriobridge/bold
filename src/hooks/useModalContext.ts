import { createContext, MutableRefObject, useContext } from 'react'
import { ModalScroll } from '../components/Modal'

export interface ModalPartsState {
  hasHeader: boolean
  hasLeftSidebar: boolean
  hasRightSidebar: boolean
}

export interface ModalContextValue extends ModalPartsState {
  scroll: ModalScroll
  bodyRef: MutableRefObject<HTMLDivElement>
  setPart: (key: keyof ModalPartsState, value: boolean) => void
  onClose?(): void
}

const ModalContext = createContext<ModalContextValue>(undefined)

export const ModalContextProvider = ModalContext.Provider

export function useModalContext() {
  const context = useContext(ModalContext)

  if (!context) {
    throw new Error(
      'Modal subcomponents (ModalHeader, ModalBody, ModalContainer, ModalFooter, ModalCloseButton) must be used inside <Modal>'
    )
  }

  return context
}
