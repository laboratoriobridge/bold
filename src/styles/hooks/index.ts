export * from './useTheme'
export * from './useCss'
export * from './useStyles'
