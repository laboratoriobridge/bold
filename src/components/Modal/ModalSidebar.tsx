import React, { useEffect } from 'react'
import { ExternalStyles, Theme, useStyles } from '../../styles'
import { useModalContext } from '../../hooks'
import { ModalScroll } from './Modal'

export type ModalSidebarPosition = 'left' | 'right'

export interface ModalSidebarProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  position: ModalSidebarPosition
  style?: ExternalStyles
}

export function ModalSidebar(props: ModalSidebarProps) {
  const { position, style, children, ...rest } = props

  const { scroll, hasHeader, setSectionState } = useModalContext()
  const { classes, css } = useStyles(createStyles, position, scroll, hasHeader)

  useEffect(() => {
    if (position === 'left') {
      setSectionState('hasLeftSidebar', true)
      return () => setSectionState('hasLeftSidebar', false)
    }
    if (position === 'right') {
      setSectionState('hasRightSidebar', true)
      return () => setSectionState('hasRightSidebar', false)
    }
  }, [setSectionState, position])

  return (
    <div className={css(classes.sidebar, style)} {...rest}>
      {children}
    </div>
  )
}

const createStyles = (theme: Theme, position: ModalSidebarPosition, scroll: ModalScroll, hasHeader: boolean) => ({
  sidebar: {
    background: theme.pallete.gray.c90,
    border: 0,
    borderLeft: position === 'right' && `1px solid ${theme.pallete.divider}`,
    borderRight: position === 'left' && `1px solid ${theme.pallete.divider}`,
    padding: '1rem',
    width: '300px',
    overflowY: scroll === 'body' ? 'auto' : 'initial',
    gridRow: hasHeader ? '2' : '1',
    gridColumn: position === 'left' ? '1' : '3',
  } as React.CSSProperties,
})
