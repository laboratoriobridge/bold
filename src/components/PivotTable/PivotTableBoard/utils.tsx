import { difference, intersection } from 'lodash'

import { ReactElement } from 'react'
import React from 'react'
import { css } from 'emotion'
import { PivotTableProps } from '../PivotTableGrid/model'
import { Tag } from '../../Tag'
import { HFlow } from '../../HFlow'
import { blue } from '../../../styles/colors'
import { Heading } from '../../Heading'
import { Icon } from '../../Icon'
import { modal } from '../../Modal'
import { LocaleConfiguration } from '../../../i18n'
import { BoardField, FieldValuesByKey } from './model'

interface InitialKeyPositionsAndValues<T> {
  initialRowKeys: Array<keyof T>
  initialColumnKeys: Array<keyof T>
  initialAvailableKeys: Array<keyof T>
}

export function getInitialKeysAndFilters<T>(
  valuesByKey: FieldValuesByKey<T>,
  allFiltersByKey: Map<keyof T, Set<string>>,
  initialFields?: Array<BoardField<T>>,
  setFilterState?: (filter: Map<keyof T, Set<string>>) => void
): InitialKeyPositionsAndValues<T> {
  const rowKeys: Array<keyof T> = []
  const columnKeys: Array<keyof T> = []
  const filtersByKey = new Map(allFiltersByKey)

  initialFields?.forEach(({ key, origin, filters }) => {
    if (origin === 'row') rowKeys.push(key)
    if (origin === 'column') columnKeys.push(key)

    if (filters.length > 0) {
      const validFilters = intersection(filters, valuesByKey.get(key))
      filtersByKey.set(key, new Set(validFilters))
    }
  })

  const availableKeys: Array<keyof T> = difference(Array.from(valuesByKey.keys()), [...rowKeys, ...columnKeys])

  setFilterState?.(filtersByKey)

  return {
    initialRowKeys: rowKeys,
    initialColumnKeys: columnKeys,
    initialAvailableKeys: availableKeys,
  }
}

export const renderClearTableModal = (labels: LocaleConfiguration['pivotTableBoard'], onTableReset: () => void) => () =>
  modal({
    size: 'small',
    render: () => (
      <Heading level={3}>
        <div css={styles.titleWrapper}>
          <Icon icon={'exclamationTriangleOutline'} fill={'danger'} size={3} style={styles.icon} />
          {labels.clearTableTitle}
        </div>
      </Heading>
    ),
    actions: [
      { label: labels.clearTableCancel },
      { label: labels.clearTableConfirm, kind: 'danger', onClick: onTableReset },
    ],
  })()

function handleTagFilterRemove<T extends object>(
  key: keyof T,
  value: string,
  filterState: Map<keyof T, Set<string>>,
  handleFilterUpdate: (key: keyof T, values: Set<string>) => void
) {
  const values = filterState.get(key) || new Set<string>()
  values.delete(value)
  handleFilterUpdate(key, values)
}

export function getFilterValuesTags<T extends object>(
  filterState: Map<keyof T, Set<string>>,
  keys: FieldValuesByKey<T>,
  keyMapping: PivotTableProps<T>['keysMapping'],
  handleFilterUpdate: (key: keyof T, values: Set<string>) => void
) {
  const filterValuesTags: ReactElement[] = []

  for (let [key, values] of Array.from(filterState)) {
    const tags: ReactElement[] = []

    if (keys.get(key)?.length === values.size || values.size === 0) {
      continue
    }

    let idx = 0
    for (let value of Array.from(values)) {
      if (idx < 3) {
        tags.push(
          <Tag
            key={value}
            removable
            onRemove={() => handleTagFilterRemove(key, value, filterState, handleFilterUpdate)}
            style={styles.tag}
          >
            {keyMapping.get(key).formatter?.(value) ?? value}
          </Tag>
        )
      } else {
        break
      }
      idx++
    }

    if (values.size > 3) {
      tags.push(
        <Tag key={key as string} type='info' style={styles.tag}>
          {`+ ${values.size - 3} ${keyMapping.get(key).keyName}`}
        </Tag>
      )
    }

    filterValuesTags.push(
      <HFlow hSpacing={0.25} alignItems='center' key={key as string}>
        <div>{`${keyMapping.get(key).keyName}`}</div>
        <div css={styles.tagsContainer}>{tags}</div>
      </HFlow>
    )
  }
  return filterValuesTags
}

const styles = {
  tag: css`
    background-color: ${blue.c90};
    color: ${blue.c40};
    border: solid 1px ${blue.c40};
    margin-left: 0.5rem;
    margin-bottom: 0.25rem;
  `,
  tagsContainer: css`
    display: flex;
    flex-wrap: wrap;
  `,
  titleWrapper: css`
    display: flex;
    align-items: center;
  `,
  icon: css`
    margin-right: 1rem;
    vertical-align: middle;
  `,
}
