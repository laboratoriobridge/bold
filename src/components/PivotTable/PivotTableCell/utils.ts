import { Theme } from '../../../styles'
import { clamp } from '../../../util'

interface CellColorProps {
  color: string
  backgroundColor: string
}

const MIN_COLOR_INDEX_OF_LIGHT_BG = 70

/**
 * Calculates text and background colors based on cell type and its content.
 *
 * If cell type is only 'value' and its content is a number:
 * - increase saturation of background color every 10% of maximum value
 * - makes the text color lighter when background becames too dark
 *
 * Otherwise, keep the default background and text color.
 *
 * @param theme
 * @param isOnlyValue True if cell type includes only 'value'
 * @param maxValue The maximum expected value for all the table
 * @param cellContent The cell's content
 * @returns An object containing the text and background colors
 */
export const calculateCellColor = (
  theme: Theme,
  isOnlyValue: boolean,
  maxValue: number,
  cellContent?: number | string
): CellColorProps => {
  let color = theme.pallete.gray.c10
  let backgroundColor = theme.pallete.primary.c100
  const cellValue = Number(cellContent)

  if (isOnlyValue && cellValue) {
    const percentageOfMaxValue = steppedPercentage(cellValue, maxValue, 10)
    const bgColorIndex = clamp(110 - percentageOfMaxValue, 10, 100)

    if (bgColorIndex < MIN_COLOR_INDEX_OF_LIGHT_BG) {
      color = theme.pallete.gray.c100
    }

    backgroundColor = theme.pallete.primary[`c${bgColorIndex}`]
  }

  return {
    color,
    backgroundColor,
  }
}

const steppedPercentage = (value: number, baseValue: number, step: number) =>
  Math.ceil(((value / baseValue) * 100) / step) * step

/**
 * Selects all the elements that has the given row and column numbers as its 'data-rownumber' and 'data-columnnumber' attributes
 */
export const selectPivotTableCellElements = (row: number, column: number): NodeListOf<HTMLElement> =>
  document.querySelectorAll(`div[data-rownumber~="${row}"], div[data-columnnumber~="${column}"]`)
