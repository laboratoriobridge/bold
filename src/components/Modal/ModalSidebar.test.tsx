import { render } from '@testing-library/react'
import React from 'react'
import { ModalContextProvider } from '../../hooks/useModalContext'
import { createMockModalContext } from '../../test/utils/createMockModalContext'
import { ModalSidebar } from './ModalSidebar'

const mockContextValue = createMockModalContext({ hasLeftSidebar: true })

it('should render correctly', () => {
  const { container } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalSidebar position='left'>Content</ModalSidebar>
    </ModalContextProvider>
  )
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalSidebar position='left' style={{ background: 'red' }}>
        Content
      </ModalSidebar>
    </ModalContextProvider>
  )
  expect(container).toMatchSnapshot()
})

it('should apply border right when side is left', () => {
  const { getByTestId } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalSidebar position='left' data-testid='sidebar-left-border'>
        Content
      </ModalSidebar>
    </ModalContextProvider>
  )
  const sidebar = getByTestId('sidebar-left-border')
  expect(getComputedStyle(sidebar).borderRight).toBe('1px solid #D3D4DD')
  expect(getComputedStyle(sidebar).borderLeft).toBe('0px')
})

it('should apply border left when side is right', () => {
  const { getByTestId } = render(
    <ModalContextProvider value={mockContextValue}>
      <ModalSidebar position='right' data-testid='sidebar-right-border'>
        Content
      </ModalSidebar>
    </ModalContextProvider>
  )
  const sidebar = getByTestId('sidebar-right-border')
  expect(getComputedStyle(sidebar).borderLeft).toBe('1px solid #D3D4DD')
  expect(getComputedStyle(sidebar).borderRight).toBe('0px')
})
