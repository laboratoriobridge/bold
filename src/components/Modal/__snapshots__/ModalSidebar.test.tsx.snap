// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  background: #F0F0F5;
  border: 0;
  border-right: 1px solid #D3D4DD;
  padding: 1rem;
  width: 300px;
  overflow-y: auto;
  grid-row: 1;
  grid-column: 1;
  background: red;
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  background: #F0F0F5;
  border: 0;
  border-right: 1px solid #D3D4DD;
  padding: 1rem;
  width: 300px;
  overflow-y: auto;
  grid-row: 1;
  grid-column: 1;
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;
